annotation-processor/src/main/java/org/hibernate/validator/ap/util/ConstraintHelper.java

```diff
@@ -544,30 +769,22 @@ else if ( assignableTypes.size() == 1 ||
 		Set<TypeMirror> supportedTypes = CollectionHelper.newHashSet();
 
 		for ( AnnotationValue oneValidatorClassReference : validatorClassReferences ) {
-
-			TypeMirror supportedType = getSupportedType( oneValidatorClassReference );
-			supportedTypes.add( supportedType );
+			
+			if(isValidationTargetSupported( oneValidatorClassReference, target )) {
+				TypeMirror supportedType = determineSupportedType( oneValidatorClassReference);
+				supportedTypes.add( supportedType );
+			}
 		}
 
 		return supportedTypes;
 	}
 
-	private TypeMirror getSupportedType(AnnotationValue oneValidatorClassReference) {
-
-		TypeMirror validatorType = oneValidatorClassReference.accept(
-				new SimpleAnnotationValueVisitor6<TypeMirror, Void>() {
-
-					@Override
-					public TypeMirror visitType(TypeMirror t, Void p) {
-						return t;
-					}
-				}, null
-		);
+	private TypeMirror determineSupportedType(AnnotationValue oneValidatorClassReference) {
 
 		// contains the bindings of the type parameters from the implemented
 		// ConstraintValidator interface, e.g. "ConstraintValidator<CheckCase, String>"
-		TypeMirror constraintValidatorImplementation = getConstraintValidatorSuperType( validatorType );
-
+		TypeMirror constraintValidatorImplementation = getConstraintValidatorSuperType( oneValidatorClassReference );
```

Comment:
not a big fan of variable names like `oneXYZ`. I think we avoided them as much as possible on the actual engine code base, but in the annotation processor we still might have more of these names left. Just saying ...